test/cpp/microbenchmarks/bm_callback_cq.cc

```diff
@@ -0,0 +1,226 @@
+/*
+ *
+ * Copyright 2015 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/* This benchmark exists to ensure that immediately-firing alarms are fast */
+
+#include <benchmark/benchmark.h>
+#include <grpc/grpc.h>
+#include <grpcpp/alarm.h>
+#include <grpcpp/completion_queue.h>
+#include <grpcpp/impl/grpc_library.h>
+#include "test/core/util/test_config.h"
+#include "test/cpp/microbenchmarks/helpers.h"
+#include "test/cpp/util/test_config.h"
+
+#include "src/core/lib/gpr/useful.h"
+#include "src/core/lib/gprpp/memory.h"
+#include "src/core/lib/iomgr/iomgr.h"
+#include "src/core/lib/surface/completion_queue.h"
+
+namespace grpc {
+namespace testing {
+
+class TagCallback : public grpc_experimental_completion_queue_functor {
+ public:
+  TagCallback(int* counter, int tag) : counter_(counter), tag_(tag) {
+    functor_run = &TagCallback::Run;
+  }
+  ~TagCallback() {}
+  static void Run(grpc_experimental_completion_queue_functor* cb, int ok) {
+    GPR_ASSERT(static_cast<bool>(ok));
+    auto* callback = static_cast<TagCallback*>(cb);
+    *callback->counter_ += callback->tag_;
+    grpc_core::Delete(callback);
+  };
+
+ private:
+  int* counter_;
+  int tag_;
+};
+
+class ShutdownCallback : public grpc_experimental_completion_queue_functor {
+ public:
+  ShutdownCallback(bool* done) : done_(done) {
+    functor_run = &ShutdownCallback::Run;
+  }
+  ~ShutdownCallback() {}
+  static void Run(grpc_experimental_completion_queue_functor* cb, int ok) {
+    *static_cast<ShutdownCallback*>(cb)->done_ = static_cast<bool>(ok);
+  }
+
+ private:
+  bool* done_;
+};
+
+/* helper for tests to shutdown correctly and tersely */
+static void shutdown_and_destroy(grpc_completion_queue* cc) {
+  grpc_completion_queue_shutdown(cc);
+  grpc_completion_queue_destroy(cc);
+}
+
+static void do_nothing_end_completion(void* arg, grpc_cq_completion* c) {}
+
+static void BM_Callback_CQ_Default_Polling(benchmark::State& state) {
+  int tag_size = state.range(0);
+  grpc_completion_queue* cc;
+  void* tags[tag_size];
```

Comment:
Variable-sized arrays are a C99 feature that aren't allowed in C++11 afaik (and not recommended by the Google C++ style guide in any case). If you want a vector, consider using grpc_core::InlinedVector . I'll also say that I'm not sure why you want the range etc for this benchmark; it's not used in the other CQ benchmarks in bm_cq.cc .